examples/fluids/navierstokes.c

```diff
@@ -897,12 +898,25 @@ int main(int argc, char **argv) {
   interpviz = NULL;
   if (viz_refine) {
     ierr = DMPlexSetRefinementUniform(dm, PETSC_TRUE); CHKERRQ(ierr);
+    dmvizfine = NULL;
     ierr = DMRefine(dm, MPI_COMM_NULL, &dmviz); CHKERRQ(ierr);
     ierr = DMSetCoarseDM(dmviz, dm); CHKERRQ(ierr);
     ierr = PetscOptionsSetValue(NULL,"-viz_petscspace_degree","1");
     CHKERRQ(ierr);
     ierr = SetUpDM(dmviz, problem, "viz_", &bc, &ctxSetup, NULL); CHKERRQ(ierr);
     ierr = DMCreateInterpolation(dm, dmviz, &interpviz, NULL); CHKERRQ(ierr);
+    for (PetscInt i = 1; i < viz_refine; i++) {
+      ierr = DMRefine(dmviz, MPI_COMM_NULL, &dmvizfine); CHKERRQ(ierr);
+      if (dmvizfine) {
+        ierr = DMSetCoarseDM(dmvizfine, dmviz); CHKERRQ(ierr);
+        ierr = PetscOptionsSetValue(NULL,"-viz_petscspace_degree","1");
+        CHKERRQ(ierr);
+        ierr = SetUpDM(dmvizfine, problem, "viz_", &bc, &ctxSetup, NULL); CHKERRQ(ierr);
+        ierr = DMCreateInterpolation(dmviz, dmvizfine, &interpviz, NULL); CHKERRQ(ierr);
```

Comment:
@knepley Shouldn't this `DmCreateInterpolation()` in the loop update the size of the interpolation matrix?

Reply:
Right but it is just interpolating between two refined levels. You have to keep all the interpolators and apply them in sequence. Also, it looks like one DM is not destroyed.

Reply:
@knepley @jedbrown 
I've been thinking how to solve this dilemma.
If I use
```
ierr = DMCreateInterpolation(dmviz, dmvizfine, &interpviz[i], NULL); CHKERRQ(ierr);
```
here with an array of length `viz_refine` of interpolation matrices, then when we [use them](https://github.com/CEED/libCEED/blob/6489aa1c689af3398a30fb46563fad9af5aa8ebd/examples/fluids/navierstokes.c#L406) in the TSMonitor_NS, I would need to have stored all the number of `viz_refine` `dm` and `dmviz` DMs, to extract all the `Q` and `Qrefined`  vectors from them of the proper size. 
I don't think this is the route we want to go in terms of memory usage.

Reply:
How much more memory do you think it is? Its probably just 2x.

Reply:
Only 2x? If I allow the user to specify, say max 10 levels of viz refinement, then it would be 10 `dm` and 10 `dmviz` DMs, from which we will extract 10 `Q` and `Qrefined` vectors, no? Am I seeing this wrong? 
I was never really concerned with memory consumption or performance before, but now I am afraid of almost anything I do in this project :sweat_smile:

Reply:
I will try it and inspect the memory usage. Thanks.

Reply:
Fixed now in this branch. @valeriabarra I'll leave it to you to squash as you see fit.

@knepley Most of the interpolator setup time is in `MatPreallocatorPreallocate_Preallocator`.

![Screenshot_20200412_204521](https://user-images.githubusercontent.com/3303/79088014-8a2d8d80-7cfe-11ea-958f-81b79fa694b8.png)

Reply:
@jedbrown just curious, how do you visualize this perf?

Reply:
http://www.brendangregg.com/FlameGraphs/cpuflamegraphs.html
```
perf record -g ./navierstokes ...
perf script | stackcollapse-perf | flamegraph > flame.svg
firefox flame.svg
```
You can click around in the SVG.

Reply:
@jedbrown It sounds like someone should speed up MatPreallocator since we are recommending people use it. I would have to do all the work it has to do in order to preallocate.

Reply:
Yeah, it could be as simple as sorting the entries so entire (block) rows could be inserted at once.

Suppose we wanted to implement refinement of Lagrange bases to the Q_1 spaces on nodes? Then the interpolation matrix would be the identity. What interface would you prefer for such a thing, or is it too special-purpose for your interests. Seems like we could change `DMPlexSetRefinementUniform` to `DMPlexSetRefinementType` with `DMPLEX_REFINE_UNIFORM, DMPLEX_REFINE_COLLOCATE, DMPLEX_REFINE_?` where collocated refinement would be unavailable for non-Lagrange bases (and any Lagrange that doesn't have natural embedded element topology).

Reply:
Let me see if I understand. You are saying that projecting P1 to Q1 is identity if the set of vertices is identical and identically numbered? That seems pretty special case. What is the generalization?

Reply:
Q_k on a coarse mesh to Q_1 on a finer mesh. It's useful for visualization so you can see the non-uniform resolution of the high-order method. This would be moot if we had good visualization for high-order bases, but that work has been slow.

Reply:
Okay, that will be a boolean matrix. I guess we could allow the user to tell us it would be boolean, which we could then use to set a floor on zeros in MatSetValues().

Reply:
We'd still have to be able to define the coordinates of the vertices on the fine mesh. I'm not sure how this would fit into your vision regarding interface, but I know the numerical code that would need to be written. I could write the code (using `DMPlexGetClosureIndices`) to preserve the ordering of vertices (so the interpolation would be a literal identity).

Reply:
I was thinking another way. We use the code exactly as it is, but tell MatSetValues() to ignore small values. Then you get the Boolean matrix. Were you thinking you could simplify the matrix construction? I was just trying to get a sparser projector.

Reply:
I could make the construction fast, but perhaps it wouldn't be a bottleneck to do naive evaluation on the interpolation nodes and discard small values. We'd still need the dispatch to get to refining to Lagrange nodes. I'd be happy to write that code (and it would be useful for preconditioning, which is reason enough to make this a well-supported functionality).